url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/render/Rasterizer2D.java

```diff
@@ -0,0 +1,458 @@
+            double xs1, xs2;
+            boolean state = false;
+
+            for ( s = 0; s < spanBuffer.size()-1; s++ ) {
+                xs1 = spanBuffer.get(s);
+                xs2 = spanBuffer.get(s+1);
+                state = state?false:true;
+
+                // Clipping...
+                if ( xs2 < minx || xs1 > maxx ) {
+                    continue;
+                }
+                else if ( xs2 < minx ) {
+                    xs2 = minx;
+                }
+                if ( xs2 > maxx ) {
+                    xs2 = maxx;
+                }
+
+                // Draw from xs1 to xs2 if "interior" state flagged
+                for ( x = (int)xs1; state == true && x < (int)xs2; x++ ) {
+                    img.putPixelRgb(x, y, color);
+                }
+            }
+
+            //-------------------------------------------------------------
+            //spanBuffer.reset();
+            spanBuffer = null;
+        }
+    }
+
+    /**
+    Given the current polygon `p` and the coordinate of a pixel (x, y),
+    this method gives the interpolated color `outColor`.
+    */
+    public static void
+    fillSmoothPolygonCalculateColor(Polygon2D p, double x, double y, ColorRgb outColor)
+    {
+        Vertex2D va = null;
+        Vertex2D vb = null;
+        int i;
+        int j;
+        double distance;
+        double totaldistance = 0.0;
+
+        outColor.r = 0.0;
+        outColor.g = 0.0;
+        outColor.b = 0.0;
+
+        for ( i = 0; i < p.loops.size(); i++ ) {
+            for ( j = 0; j < p.loops.get(i).vertices.size(); j++ ) {
+                va = p.loops.get(i).vertices.get(j);
+                distance = 1.0/(1.0+Math.sqrt((va.x-x)*(va.x-x) + 
+                                              (va.y-y)*(va.y-y)));
+                totaldistance += distance;
+                outColor.r += va.color.r * distance;
+                outColor.g += va.color.g * distance;
+                outColor.b += va.color.b * distance;
+            }
+        }
+
+        outColor.r /= totaldistance;
+        outColor.g /= totaldistance;
+        outColor.b /= totaldistance;
+    }
+
+    /**
+    Current polygon filling rasterizer (scan-line) algorithm is a NAIVE
+    implementation of the general macro-algorithm outlined at [FOLE1992].3.6.
+
+    This implementation is working but inefficient, due to the fact that
+    neither scanline coherence nor edge coherence is taken into account. This
+    means that for each scanline, all polygon edges are intersected 
+    analitically. This is the "brute-force technique" recommended to be
+    avoided in [FOLE1992].3.6.3, but it is provided as reference to
+    compare efficiency with other (clever) aproaches that makes use of
+    mid-point algoritms for intersection finding, or "active edge tables"
+    (AETs) for efficient edge-coherence based traversals.
+    */
+    public static void fillSmoothPolygon(Image img, Polygon2D p)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 28. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.